python/ray/autoscaler/v2/scheduler.py

```diff
@@ -242,3 +247,278 @@ def __repr__(self) -> str:
+                self._nodes, self._cluster_config
+            )
+
+        def get_cluster_config(self) -> ClusterConfig:
+            return self._cluster_config
+
+        def __str__(self) -> str:
+            return "ScheduleContext({} nodes, node_type_available={}): {}".format(
+                len(self._nodes), self._node_type_available, self._nodes
+            )
+
+    def schedule(self, request: SchedulingRequest) -> SchedulingReply:
+        self._init_context(request)
+
+        # 1. Enforce the minimal count of nodes for each worker node type.
+        self._enforce_min_workers()
+
+        # 2. Enforce the cluster resource constraints.
+        infeasible_constraints = self._enforce_resource_constraints(
+            request.cluster_resource_constraints
+        )
+
+        # 3. Schedule the gang resource requests.
+        infeasible_gang_requests = self._sched_gang_resource_requests(
+            request.gang_resource_requests
+        )
+
+        # 4. Schedule the tasks/actor resource requests
+        infeasible_requests = self._sched_resource_requests(
+            request.resource_requests,
+        )
+
+        # Compute the number of nodes to launch.
+        reply = SchedulingReply(
+            infeasible_resource_requests=resource_requests_by_count(
+                infeasible_requests
+            ),
+            infeasible_gang_resource_requests=infeasible_gang_requests,
+            infeasible_cluster_resource_constraints=infeasible_constraints,
+            target_cluster_shape=self._ctx.get_cluster_shape(),
+        )
+
+        return reply
+
+    def _init_context(self, request: SchedulingRequest) -> None:
+        self._ctx = self.ScheduleContext.from_schedule_request(request)
+
+    def _enforce_min_workers(self) -> None:
+        """
+        Enforce the minimal count of nodes for each worker node type.
+
+        Args:
+            config: The cluster config.
+
+        """
+
+        # Count the existing nodes by type
+        count_by_node_type = self._ctx.get_cluster_shape()
+        logger.debug("Enforcing min workers: {}".format(self._ctx))
+
+        new_nodes = []
+        # Launch new nodes to satisfy min count for each node type.
+        for (
+            node_type,
+            node_type_config,
+        ) in self._ctx.get_cluster_config().node_type_configs.items():
+            cur_count = count_by_node_type.get(node_type, 0)
+            min_count = node_type_config.min_workers
+            if cur_count < min_count:
+                new_nodes.extend(
+                    [
+                        SchedulingNode.from_node_config(
+                            copy.deepcopy(node_type_config),
+                            status=SchedulingNodeStatus.TO_LAUNCH,
+                        )
+                    ]
+                    * (min_count - cur_count)
+                )
+        # NOTE: we assume the aggregated number of min workers across all node types
```

Comment:
Do we have validation somewhere to ensure this?

Reply:
this should be something our config checks.